doc/tutorial/tutorial.md

```diff
@@ -987,6 +987,34 @@ BestPricePerVendor(item, vendor, best_price) :-
     var best_price = Aggregate((item, vendor), group_min(price)).
 ```
 
+The aggregation function can take group-by variables in addition to
+the group.  The following custom aggregation function computes the cheapest
+vendor for each item and returns a string containing item name, vendor,
+and price:
+
+```
+function best_vendor_string(item: string, g: Group<(string, bit<64>)>): string =
```

Comment:
Frankly I find this quite confusing. The signature of the function is not the same in the declaration and in the call.
Is backwards compatibility so important? If you want to make changes to the language you should do them while it still has few users. You can provide a good diagnostic to help people migrate their code. 

I think it's cleaner if the Group object just has a key field. Group should be templated on two types, the key type and the value type. If you really want backwards compatibility you can have the for operate directly on the group. In LINQ this is done by the group implementing the `IEnumerable<Value>` interface.

Reply:
Let's run this by Ben, he has very good taste in software design.